pyomo/core/base/constraint.py

```diff
@@ -331,6 +331,58 @@ def __init__(self,  expr=None, component=None):
         if expr is not None:
             self.set_value(expr)
 
+        self._canonical_form_has_been_set = False
+        self._canonical_form = None
+
+    @property
+    def _linear_canonical_form(self):
+        return self._canonical_form_has_been_set or _ConstraintData._linear_canonical_form
+
+    def cast_to_linear_canonical_form(self):
```

Comment:
To be called by the external calling environment when the model is ready to be "finalised"